src/python/pants/backend/python/macros/python_requirements_new.py

```diff
@@ -0,0 +1,131 @@
+from pants.engine.addresses import Address
+from pants.engine.collection import Collection
+from pants.engine.fs import DigestContents, GlobMatchErrorBehavior, PathGlobs
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule, rule
+from pants.engine.target import Target, UnexpandedTargets, WrappedTarget
+from pants.engine.unions import UnionMembership, UnionRule, union
+
+logger = logging.getLogger(__name__)
+
+
+_T = TypeVar("_T", bound=Target)
+
+
+@union
+@dataclass(frozen=True)
+class GenerateTargetsRequest:
+    target_class: ClassVar[type[_T]]
+    target: _T
+
+
+class GeneratePythonRequirementLibrariesFromPythonRequirements(GenerateTargetsRequest):
+    target_class = PythonRequirementsMacro
+
+
+class GeneratedTargets(Collection[Target]):
+    pass
+
+
+@rule
+async def generate_python_requirement_libraries_from_requirements_file(
+    request: GeneratePythonRequirementLibrariesFromPythonRequirements,
+) -> GeneratedTargets:
+    relpath = os.path.join(
+        request.target.address.spec_path, request.target[PythonRequirementsRelpath].value
+    )
+    digest_contents = await Get(
+        DigestContents,
+        PathGlobs(
+            [relpath],
+            glob_match_error_behavior=GlobMatchErrorBehavior.error,
+            description_of_origin=f"{request.target}'s field `{PythonRequirementsRelpath.alias}`",
+        ),
+    )
+    requirements = parse_requirements_file(digest_contents[0].content.decode(), rel_path=relpath)
+    grouped_requirements = itertools.groupby(
+        requirements, lambda parsed_req: parsed_req.project_name
+    )
+    result = []
+    module_mapping = {}
+    type_stubs_module_mapping = {}
+    for project_name, parsed_reqs in grouped_requirements:
+        req_module_mapping = (
+            {project_name: module_mapping[project_name]}
+            if module_mapping and project_name in module_mapping
+            else None
+        )
+        stubs_module_mapping = (
+            {project_name: type_stubs_module_mapping[project_name]}
+            if type_stubs_module_mapping and project_name in type_stubs_module_mapping
+            else None
+        )
+        generated_tgt = PythonRequirementLibrary(
+            {
+                PythonRequirementsField.alias: list(parsed_reqs),
+                ModuleMappingField.alias: req_module_mapping,
+                TypeStubsModuleMappingField.alias: stubs_module_mapping,
+            },
+            Address(request.target.address.spec_path, target_name=project_name),
+        )
+        result.append(generated_tgt)
+
+    return GeneratedTargets(result)
+
+
+class TargetGenSubsystem(GoalSubsystem):
+    name = "target-gen"
+    help = "Foo"
+    required_union_implementations = (GenerateTargetsRequest,)
+
+
+class TargetGen(Goal):
+    subsystem_cls = TargetGenSubsystem
+
+
+@goal_rule
+async def gen_targets(union_membership: UnionMembership) -> TargetGen:
```

Comment:
> Target generation generates 0-n targets from a single target. That base target uses the Target API like normal, meaning that python_requirements will no longer be some special type but just a normal target.

This comment made me realize that this has a lot in common with `Subtargets`, as they exist today.

Rather than a global scan, if we're comfortable leaning in on "there must be an anchor target in the BUILD file" (...which seems pretty reasonable?), then this could probably be hooked in as an alternative-`Subtarget` generation scheme.

The main complexity there is that the `Subtargets` of a `Target` are currently keyed off of their `Addresses`, and so you'd need an address that referred to a `Target` that didn't actually live on disk. It might mean adjusting the meaning of the "file component" of the address to mean something more generic ("child component"? "sub component"?)

But... that seems like an issue that this PR would need to address as well in order to be useful.

Reply:
True! Good point. I could definitely see us replacing Subtargets with this.

And yes, indeed, these generated targets will need to be addressable. That will be the tricky part: `./pants list ::` needs to show all targets, for example.

Reply:
This fantastic insight resulted in https://docs.google.com/document/d/1HpJn2jTWf5sKob6zhe4SqHZ7KWBlX4a6OJOwnuw-IHo/edit#heading=h.7dq13i2i8pzd :)